### PR TITLE
[WIP] Faster picking for ArcGIS Terrain / Heightmaps

### LINKED ISSUE #8481: Globe picking is very slow
The current approach for globe picking is intersect the ray against all rendered tiles (using their bounding spheres) and sort the intersected tiles by closeness to ray. Then, for the closest tile do ray-triangle intersection tests against all of its triangles and return the closest. If there was none, test the next tile. And repeat.

This could be optimized some more. `ScreenSpaceCameraController` does one or two globe picks per frame and that alone can slow down Cesium severely with detailed terrain providers like `ArcGIS`.

Some ideas on how to improve this:

- Custom intersection test for heightmap providers since we know their triangles are on a grid
- Per-tile octree for non-heightmap providers (too slow to generate?)
- Remove `TerrainEncoding` quantization for faster position decode from vertex buffer (only affects nearby tiles)
- Cache recently used triangles in a form that makes future picks faster, maybe by pre-calculating some of the math from `IntersectionTests.rayTriangleParametric`.
- Unroll picking math

[Sandcastle](https://sandcastle.cesium.com/#c=hVXdbupGEH6VFTcxElnbmNgmIVGPOLQHnaSNgnWqqlTVxp7A6phda3cNoVXu+mZ9sY7/wEBofGPv/Hwz8+3MeM0UgTTlmZY8eVRyzRNQ5JYI2JAxaJ6v6KRRR6AU46Kxsro3czEXa0RYq2gJv0qVnoH49tTojzH+nguCT65Sck0ulsZk17a92WzoWl2aJWwKHxrLld0624anoF3qUMcOPPuiV2HEChJuCpg6CEmYYSSWuTKgt+RbRB7+/efrxVy87TNnKv6J66OsDpP/hCbTWYQxk0kKa2a4FOfLqKvQWAY01l5CMc6C67KS6tNWoI2tQa15DNouydnBe5/tOgJ+TVdsATM0BHWYe1zm16YVE6+TRjKYOaC8fV0cNidXVMqsiwp0LIVBH4zYI3Vt5oij63fi79Mr0Si8GhCJVYeohNVhKnQGsZHqgb9ysU9NxyAAM6sByuNNXS5bgWKoK4W0Ota6RSqfYacqT7Vm19yoLeV0JymDvuQiLignesNNvGzIOqq227CgwORKkMbL2imKp07anHTTkeSmcnnD91tTuMwKPI3Gv+8Ify27uSKMlDyTOr+m59EDuywu7A7zP72bLgZrXWaFvR/MDxFPZ/xdxGpaPkR7d+7eBdytnw8xT/ZYgfdHecszJpKYaZMCZUkSSZk+M/UAIrdq3g/att1gVIMpZqMZ8QTHlotySjGB1gCNWbFmOBOedekO3eFgGNArz3FDZ+i5PXI5CPDL92kYFu8g6PeIF1zhZ0gdJwyGV77nDUi3KVJxwBE8DfMFWMLF4rEo/EmmqeXSvn/lDjw/6AdePwzCAQZzCqHnu8PQd/u+5zphj+C6dDzX7/to5zthHy3dJhxOaaSY0C9SrTBeHeuBGcVfB3T6efJzNI1+O9g+S6Q0Pd4iM9w7IGYZi2GyxgK+VEbWEbEIUDJeYxQcT0WWm0/VUO2mayXXsEKY3ZT9z4po1OfWxN5iJXMNj1Lzkt1b0kShWS1roSm2RYu6FRZgHnn8/YltrQOM7gH8MyCLRYIZqIJPJmKgQm7KBdwYZYjTSqG1t2ihsjBur5K2vdiLKRk/hxxjJ0vs8VQurMr2sk6nsMDZOnNN0TYDej/5MfpzfD8dfy2MO73OSJttCncV9A98lUllil+cRaltYJWl+IfR9nMefwdDY13e6MhunEYJXxOe3M47R7+UeYfEKdMaNS95ms74XzDv3I1stD9wS2XZ6b/gfy9l28Jk6d7dV0JK6cjG46mXqWa7hfgf) (console prints how long each pick takes with left click). Select `ArcGIS` from the dropdown for some seriously slow speeds.

[WIP branch](https://github.com/CesiumGS/cesium/tree/fastPicking/)

### PR DESCRIPTION
**Pull request by DanielLeone**
Hi, this PR fixes #8481 but only for height maps.

It's still a work in progress, so there's a few bugs I would like some help ironing out if possible.
But for the most part it's all working :)

@IanLilleyT has helped a considerable amount with this too - thanks!

## What is this for? 
An ArcGIS Tiled Elevation Service is very slow in Cesium currently. Reason being, the simple implementation of globe picking just goes through every triangle in the mesh... and height maps have way more triangles than a CWT tile for example. 

dat frame rate
image

## How it's been fixed

A quadtree.

1. Take each tile - and make a bounding box for it.
image

2. Divide that bounding box into 4 quadrants on the horizontal plane and for each segment find the minimum and maximum heights.
image

3. Repeat to a desired quadtree depth (currently 5).

image

4. When "picking" a ray against a tile:
    a. Quickly traverse the quadtree using the bounding boxes
    b. This effectively narrows down the number of triangles/area we need to test
    c. At the bottom of the quadtree - test all triangles within the bounding boxes which intersected with the ray

The red boxes here at the ones intersecting with the ray:
image

The orange patches on the terrain (under the red boxes) are the triangles we actually ended up testing against the ray

image

This is much faster than testing every triangle for the whole tile.

## Remaining issues
1. The system falls apart at very high zoom levels; and at the poles.

Exhibit A - strange heights on the outer layer of green boxes:
image

Exhibit B - green boxes start to distort from the evenly divided grid.
image

2. I think I'd like to rotate the transform matrix (from world space to the local tile space) so it's orientated the same as the height map data itself (North to South, West to East, I believe). Currently it always seems to come out East to West, North to South; so I'm just doing some funky math to "rotate the matrix" as you look up a position in it (confusing). I did attempt this but it didn't go well... 

3. I haven't thought about 2D mode at all - what are the considerations/differences with 2D mode?

## What do you want?
If someone has some time to help me out - that'd be much appreciated.
You can have a play in the Sandcastle - I've hijacked `Terrain.html`
* left click to "debug" a ray pick; change with the dropdowns in the top left to see the things
image

## Notes
* There's also some Octree code in here - that was a prior implementation - don't worry about that one... it's all quadtrees now.
* There's also fixes from #9286  in here (because it hasn't been merged yet)
* It's highly possible I've made an error somewhere and then covered it up with a equivalent compensating error elsewhere to make everything work - I wouldn't put it past me.

Much thanks!

### CODE REVIEW

**Comment by cesium-concierge** [comment]
Thank you so much for the pull request @DanielLeone! I noticed this is your first pull request and I wanted to say welcome to the Cesium community!

The [Pull Request Guidelines](https://github.com/CesiumGS/cesium/blob/master/CONTRIBUTING.md#pull-request-guidelines) is a handy reference for making sure your PR gets accepted quickly, so make sure to skim that.

* :x: Missing `CONTRIBUTORS.md` entry.
    * Please add yourself to the [contributors](https://github.com/DanielLeone/cesium/blob/fastPicking2/CONTRIBUTORS.md) file!
* :heavy_check_mark: Signed CLA found.
* :grey_question: Unit tests were not updated.
    * Make sure you've [updated tests](https://github.com/CesiumGS/cesium/tree/master/Documentation/Contributors/TestingGuide) to reflect your changes, added tests for any new code, and ran the code coverage tool.

Reviewers, don't forget to make sure that:

- [ ] Cesium Viewer works.
- [ ] Works in 2D/CV.
- [ ] Works (or fails gracefully) in IE11.

File: Source/Core/HeightmapTessellator.js [hunk 1]
```diff
@@ -452,27 +601,79 @@ HeightmapTessellator.computeVertices = function (options) {
...
+      0,             0,   1,  0,
+      0,             0,   0,  1
+    );
+    // transform = Matrix4.multiply(transform, zRotation, transform);
+    inverseTransform = Matrix4.inverse(transform, new Matrix4());
+
+    console.timeEnd("creating oriented bounding box");
+
+    console.time("making packed quadtree");
+    quadtree = createPackedQuadtree(
```
Comment (DanielLeone) [inline Source/Core/HeightmapTessellator.js:655]: creation of the quad tree lives here

File: Source/Core/QuadtreeTrianglePicker.js [hunk 2]
```diff
@@ -0,0 +1,388 @@
...
+  };
+};
+
+/**
+ * @param {Ray} ray
+ * @param {Boolean} cullBackFaces
+ * @param {Cartesian3} result
+ * @returns {Cartesian3} result
+ */
+QuadtreeTrianglePicker.prototype.rayIntersect = function (
```
Comment (DanielLeone) [inline Source/Core/QuadtreeTrianglePicker.js:232]: algorithm to pick a ray using the quadtree starts here

**Comment by IanLilleyT** [comment]
I looked around a bit but didn't go very deep, partly because I kept running into an error when picking from a high zoom level, so I couldn't repro the misaligned boxes:

image

One possibly unrelated issue is that when the quadtree is created, the highest vertex that gets added to a node is not necessarily the highest extent of the node, since a tall vertex may have an even taller neighbor that is part of a different node, so the node's maximum extent falls somewhere along the triangle between them. This wouldn't be a problem if the local space was defined by lat / lon / height (which is a non-starter anyway because it would slow down node intersection tests), but since it's defined by the OBB transformation things don't line up as well. I don't think this explains the misaligned boxes though...

image

In terms of the API for all of this, I've been playing around with the idea of adding a `pick` function to `TerrainData`. This way we wouldn't have to modify `TerrainMesh` at all and each `TerrainData` could implement its own picking algorithms. For example, future versions of Cesium's TerrainProvider may have acceleration structures as part of the tile payload. Since the `HeightmapTerrainData` acceleration is basically just a dense grid of boxes, would could serialize as one big node height array and get rid of node objects altogether.

**Comment by DanielLeone** [comment]
Hi @IanLilleyT thanks for taking a look; sorry for the errors you ran into.

In regards to the issue error you pointed out - good catch :) I think this is ok for now - I'll just have to make the height calculation of each quadtree node take into account "an extra layer" of positions for the min-max height.. I think that's fine - just something else to do 👍 .

API design wise - I think I had a similar idea; currently the `TerrainData` class provides a pick 'strategy' to the `TerrainMesh` constructor. Can obviously shuffle all this later 👍 

I've got this picture here I hope can help describe the issue I'm having.
image


**Cyan "rectangle"**: The "rectangle" provided directly by the ArcGIS terrain tile (so the positions/vertices/heightmap points reside within in this rectangle).

**Yellow "rectangle"**: This is meant to be`[(-0.5, -0.5), (0.5, 0.5)]` from local space transformed to world space and drawn as a rectangle... Obviously I'm doing something wrong here as it doesn't line up with anything here; but it's not the main concern right now I think.


**All the little blue/yellow/green/orange points**: These are various points spread around `(-0.5,-0.5,-0.5)` to `(0.5,0.5,0.5)` at increments of 0.1 ... so this is my best representation of the "Oriented Bounding Box". This seems correct to me as it does make a "bounding box" around the rectangle at a non-specific orientation.

image


I think this is where my problem/understanding is though.
I guess I was hoping that `-0.5,-0.5` in "local" space corresponded to the top left corner of the cyan rectangle.
Or put another way; I hoped all the blue/green/yellow/orange puts would neatly fit inside the cyan rectangle.
Or put another way; I hoped the "transform" that's created from the bounding box - "exactly fit" the cyan rectangle.

Obviously that's not the case :)

(unless you're closer to the equator and at a higher zoom level - then every does begin to "line up" more and more)

image

So...
1. Does that summary of what's-what make sense to you? 
2. Does Cesium have a ray-box intersection test already? (I couldn't find one) so that we can do this all in local space (ie divide up the cyan rectangle evenly for the quadtree).

I know you said:

> which is a non-starter anyway because it would slow down node intersection tests

but I'm initially skeptical - we don't need many intersections test of the quadtree boxes - I think it's worth testing - especially given I'm out of other ideas 😄 .

Thanks for the help 👍

**Comment by IanLilleyT** [comment]
Thanks for the extra information and visuals. Wanted to let you know that unfortunately I am kind of swamped right now so it could be a couple weeks until I can dig into this some more.

**Comment by DanielLeone** [comment]
> Thanks for the extra information and visuals. Wanted to let you know that unfortunately I am kind of swamped right now so it could be a couple weeks until I can dig into this some more.

No stress 👍 thanks for the update.
I'm still thinking about this as well - but I can see why this isn't immediately going to work as in.

**Comment by cesium-concierge** [comment]
Thanks again for your contribution @DanielLeone! 

No one has commented on this pull request in 90 days. Maintainers, can you review, merge or close to keep things tidy?

 _I'm going to re-bump this in 90 days. If you'd like me to stop, just comment with `@cesium-concierge stop`. If you want me to start again, just delete the comment._

**Comment by cesium-concierge** [comment]
Thanks again for your contribution @DanielLeone! 

No one has commented on this pull request in 90 days. Maintainers, can you review, merge or close to keep things tidy?

 _I'm going to re-bump this in 90 days. If you'd like me to stop, just comment with `@cesium-concierge stop`. If you want me to start again, just delete the comment._

**Comment by DanielLeone** [comment]
I'm going to close this PR for now - but it's still in my todo list :) I've gone back to an octree based algorithm, but using a different implementation :)

**Pull request closed without merging**